tools/toolchains/iar.py

```diff
@@ -154,8 +154,10 @@ def get_compile_options(self, defines, includes, for_asm=False):
 
     @hook_tool
     def assemble(self, source, object, includes):
+        _, macros = self.config.get_config_data()
+        defines = ['-D%s' % d for d in macros] if macros else [""]
```

Comment:
I'm pretty sure this is the same as the result of `self.get_compile_options(self.get_symbols(True), includes, True)` below, so the changes in this file are not needed.

Reply:
No, I tried this at first, but the macros from the mbed_app.json and mbed_lib.json files are not included in the set returned by self.get_compile_options(self.get_symbols(True), includes, True). If they are supposed to, then perhaps this is a bug?

Reply:
`get_symbols` is a shared bit of infrastructure so we can eliminate any compiler differences caused by it. This leaves `get_compile_options`
arm: https://github.com/ARMmbed/mbed-os/blob/master/tools/toolchains/arm.py#L136-L139
gcc: https://github.com/ARMmbed/mbed-os/blob/master/tools/toolchains/gcc.py#L144-L155

They seem to disagree on whether or not to include the include paths, but they both treat the defines passed in the same.

This makes me think that the issue this is trying to fix will introduce an inconsistency between the assemblers. That's something we should avoid.

Reply:
I'm happy to have it fixed globally in get_symbols(), if that is doable.

Reply:
That's doable, and I think that's how it should be done.